File: SJSegmentedScrollView-ObjC/SJSegmentedScrollView.m
//
//  SJSegmentedScrollView.m
//  Hike
//
//  Created by Pavan Goyal on 02/05/18.
//  Copyright © 2018 Hike Pvt. Ltd. All rights reserved.
//

#import "SJSegmentedScrollView.h"

@interface SJSegmentedScrollView ()

@end

@implementation SJSegmentedScrollView

- (instancetype)initWithFrame:(CGRect)frame {
    self = [super initWithFrame:frame];
    if (self) {
        _headerViewHeight = 0.0;
        _segmentViewHeight = 0.0;
        _headerViewOffsetHeight = 0.0;
        _selectedSegmentViewColor = [UIColor redColor];
        _selectedSegmentViewHeight = 0.0;
        _segmentBounces = NO;
        _segmentTitleColor = [UIColor redColor];
        _segmentViewBackgroundColor = [UIColor grayColor];
        _segmentTitleFont = [UIFont systemFontOfSize:12];
        _topSpacing = 0.0;
        _bottomSpacing = 0.0;
        _observing = YES;
        _contentControllers = [[NSMutableArray alloc] init];
        _contentViews = [[NSMutableArray alloc] init];
        _sjShowsVerticalScrollIndicator = NO;
        _sjShowsHorizontalScrollIndicator = NO;
        _viewObservers = [[NSMutableArray alloc] init];
        _sjDisableScrollOnContentView = NO;
        [self sizeToFit];
        self.translatesAutoresizingMaskIntoConstraints = NO;
        self.showsHorizontalScrollIndicator = NO;
        self.showsVerticalScrollIndicator = NO;
        self.decelerationRate = UIScrollViewDecelerationRateFast;
        self.bounces = NO;
        [self addObserver:self forKeyPath:@"contentOffset" options:NSKeyValueObservingOptionNew|NSKeyValueObservingOptionOld context:nil];
    }
    return self;
}

- (void)setSjShowsVerticalScrollIndicator:(BOOL)sjShowsVerticalScrollIndicator {
    _sjShowsVerticalScrollIndicator = sjShowsVerticalScrollIndicator;
    self.showsVerticalScrollIndicator = sjShowsVerticalScrollIndicator;
    self.contentView.showsVerticalScrollIndicator = sjShowsVerticalScrollIndicator;
}

- (void)setSjShowsHorizontalScrollIndicator:(BOOL)sjShowsHorizontalScrollIndicator {
    _sjShowsHorizontalScrollIndicator = sjShowsHorizontalScrollIndicator;
    self.showsHorizontalScrollIndicator = sjShowsHorizontalScrollIndicator;
    self.contentView.showsHorizontalScrollIndicator = sjShowsHorizontalScrollIndicator;
}

- (void)setSjDisableScrollOnContentView:(BOOL)sjDisableScrollOnContentView {
    _sjDisableScrollOnContentView = sjDisableScrollOnContentView;
    [self.contentView setScrollEnabled:!sjDisableScrollOnContentView];
}

- (void)dealloc {
    [self removeObserver:self forKeyPath:@"contentOffset" context:nil];
    for (UIView *view in self.viewObservers) {
        [view removeObserver:self forKeyPath:@"contentOffset" context:nil];
    }
}

- (void)setContentView {
    if (self.scrollContentView == nil) {
        self.scrollContentView = [[UIView alloc] init];
        self.scrollContentView.translatesAutoresizingMaskIntoConstraints = NO;
        [self addSubview:self.scrollContentView];
        NSArray<NSLayoutConstraint *> *horizontalConstraints = [NSLayoutConstraint constraintsWithVisualFormat:@"H:|[contentView(==mainView)]|" options:0 metrics:nil views:@{@"contentView": self.scrollContentView, @"mainView": self}];
        [self addConstraints:horizontalConstraints];
        CGFloat contentHeight = [self getContentHeight];
        NSArray<NSLayoutConstraint *> *verticalConstraints = [NSLayoutConstraint constraintsWithVisualFormat:@"V:|[contentView]|" options:0 metrics:nil views:@{@"contentView": self.scrollContentView}];
        [self addConstraints:verticalConstraints];
        self.contentViewHeightConstraint = [NSLayoutConstraint constraintWithItem:self.scrollContentView attribute:NSLayoutAttributeHeight relatedBy:NSLayoutRelationEqual toItem:nil attribute:NSLayoutAttributeNotAnAttribute multiplier:1.0 constant:contentHeight];
        [self addConstraint:self.contentViewHeightConstraint];
    }
}

- (void)addHeaderView:(UIView *)headerView {
    if (headerView != nil) {
        self.headerView = headerView;
        headerView.translatesAutoresizingMaskIntoConstraints = NO;
        [self.scrollContentView addSubview:headerView];
        NSArray<NSLayoutConstraint *> *horizontalConstraints = [NSLayoutConstraint constraintsWithVisualFormat:@"H:|-0-[headerView]-0-|" options:0 metrics:nil views:@{@"headerView": headerView}];
        [self.scrollContentView addConstraints:horizontalConstraints];
        NSArray<NSLayoutConstraint *> *verticalConstraints = [NSLayoutConstraint constraintsWithVisualFormat:@"V:|[headerView(headerViewHeight)]" options:0 metrics:@{@"headerViewHeight": @(self.headerViewHeight)} views:@{@"headerView": headerView}];
        [self.scrollContentView addConstraints:verticalConstraints];
    } else {
        self.headerViewHeight = self.headerViewOffsetHeight;
    }
}

- (void)addObserverForView:(UIView *)view {
    [self.viewObservers addObject:view];
    [view addObserver:self forKeyPath:@"contentOffset" options:NSKeyValueObservingOptionNew|NSKeyValueObservingOptionOld context:nil];
}

- (void)addContentView:(UIView *)contentView frame:(CGRect)frame {
    if (self.contentView == nil) {
        self.contentView = [self createContentView];
    }
    [self.contentViews addObject:contentView];
    [self.contentView addContentView:contentView frame:frame];
    __weak typeof(self) weakSelf = self;
    self.contentView.didSelectSegmentAtIndex = ^(SJSegmentTab *segmentTab, NSInteger index, BOOL animated) {
        SJSegmentTab *newSegmentTab = [weakSelf.segmentView.segments objectAtIndex:index];
        if ([newSegmentTab isKindOfClass:[SJSegmentTab class]]) {
            if (weakSelf.didSelectSegmentAtIndex) {
                weakSelf.didSelectSegmentAtIndex(newSegmentTab, index, animated);
            }
        }
    };
}

- (void)updateSubviewsFrame:(CGRect)frame {
    self.contentViewHeightConstraint.constant = [self getContentHeight];
    [self.contentView layoutIfNeeded];
    [self.segmentView didChangeParentViewFrame:frame];
    [self.contentView updateContentControllersFrame:frame];
}

- (CGFloat)getContentHeight {
    CGFloat contentHeight = self.superview.bounds.size.height + self.headerViewHeight;
    contentHeight -= (self.topSpacing + self.bottomSpacing + self.headerViewOffsetHeight);
    return contentHeight;
}

- (void)addSegmentView:(NSArray<UIViewController *> *)controllers frame:(CGRect)frame {
    if (controllers.count > 1) {
        self.segmentView = [[SJSegmentView alloc] initWithFrame:CGRectZero];
        self.segmentView.backgroundColor = self.segmentViewBackgroundColor;
        self.segmentView.controllers     = controllers;
        self.segmentView.selectedSegmentViewColor        = self.selectedSegmentViewColor;
        self.segmentView.selectedSegmentViewHeight       = self.selectedSegmentViewHeight;
        self.segmentView.titleColor                      = self.segmentTitleColor;
        self.segmentView.segmentBackgroundColor          = self.segmentBackgroundColor;
        self.segmentView.font                            = self.segmentTitleFont;
        self.segmentView.font                            = self.segmentTitleFont;
        self.segmentView.segmentViewHeight               = self.segmentViewHeight;
        self.segmentView.bounces                         = NO;
        self.segmentView.translatesAutoresizingMaskIntoConstraints = NO;
        __weak typeof(self) weakSelf = self;
        self.segmentView.didSelectSegmentAtIndex = ^(SJSegmentTab *segmentTab, NSInteger index, BOOL animated) {
            [weakSelf.contentView movePageToIndex:index animated:animated];
            if (weakSelf.didSelectSegmentAtIndex) {
                weakSelf.didSelectSegmentAtIndex(segmentTab, index, animated);
            }
        };
        [self.segmentView setSegmentsView:frame];
        [self addSubview:self.segmentView];
        NSArray<NSLayoutConstraint *> *horizontalConstraints = [NSLayoutConstraint constraintsWithVisualFormat:@"H:|-0-[segmentView]-0-|" options:0 metrics:nil views:@{@"segmentView": self.segmentView}];
        [self addConstraints:horizontalConstraints];
        UIView *view = self.headerView == nil ? self : self.headerView;
        NSArray<NSLayoutConstraint *> *verticalConstraints = [NSLayoutConstraint constraintsWithVisualFormat:@"V:[headerView]-0-[segmentView(segmentViewHeight)]" options:0 metrics:@{@"segmentViewHeight" : @(self.segmentViewHeight)} views:@{@"headerView": view, @"segmentView": self.segmentView}];
        [self addConstraints:verticalConstraints];
    } else {
        self.segmentViewHeight = 0.0;
    }
}

- (void)addSegmentsForContentViews:(NSArray<NSString *> *)titles {
    CGRect frame = CGRectMake(0, self.headerViewHeight, self.bounds.size.width, self.segmentViewHeight);
    self.segmentView = [[SJSegmentView alloc] initWithFrame:frame];
    __weak typeof(self) weakSelf = self;
    self.segmentView.didSelectSegmentAtIndex = ^(SJSegmentTab *segmentTab, NSInteger index, BOOL animated) {
        [weakSelf.contentView movePageToIndex:index animated:animated];
    };
    [self addSubview:self.segmentView];
}

- (SJContentView *)createContentView {
    SJContentView *contentView = [[SJContentView alloc] initWithFrame:CGRectZero];
    contentView.showsVerticalScrollIndicator = self.sjShowsVerticalScrollIndicator;
    contentView.showsHorizontalScrollIndicator = self.sjShowsHorizontalScrollIndicator;
    [contentView setScrollEnabled:!self.sjDisableScrollOnContentView];
    contentView.translatesAutoresizingMaskIntoConstraints = NO;
    contentView.bounces = self.segmentBounces;
    [self.scrollContentView addSubview:contentView];
    NSArray<NSLayoutConstraint *> *horizontalConstraints = [NSLayoutConstraint constraintsWithVisualFormat:@"H:|-0-[contentView]-0-|" options:0 metrics:nil views:@{@"contentView": contentView}];
    [self.scrollContentView addConstraints:horizontalConstraints];
    NSArray<NSLayoutConstraint *> *verticalConstraints = [NSLayoutConstraint constraintsWithVisualFormat:@"V:[headerView]-segmentViewHeight-[contentView]-0-|" options:0 metrics:@{@"segmentViewHeight": @(self.segmentViewHeight)} views:@{@"headerView": self.headerView, @"contentView": contentView}];
    [self.scrollContentView addConstraints:verticalConstraints];
    return contentView;
}

- (void)handleScrollUp:(UIScrollView *)scrollView change:(CGFloat)change oldPosition:(CGPoint)oldPosition {
    if (self.headerViewHeight != 0.0 && self.contentOffset.y != 0.0) {
        if (scrollView.contentOffset.y < 0.0) {
            if (self.contentOffset.y >= 0.0) {
                CGFloat yPos = self.contentOffset.y - change;
                yPos = (yPos < 0) ? 0 : yPos;
                CGPoint updatedPos = CGPointMake(self.contentOffset.x, yPos);
                [self setContentOffset:self point:updatedPos];
                [self setContentOffset:scrollView point:oldPosition];
            }
        }
    }
}

- (void)handleScrollDown:(UIScrollView *)scrollView change:(CGFloat)change oldPosition:(CGPoint)oldPosition {
    CGFloat offset = (self.headerViewHeight - self.headerViewOffsetHeight);
    if (self.contentOffset.y < offset) {
        if (scrollView.contentOffset.y >= 0.0) {
            CGFloat yPos = self.contentOffset.y - change;
            yPos = (yPos > offset) ? offset : yPos;
            CGPoint updatedPos = CGPointMake(self.contentOffset.x, yPos);
            [self setContentOffset:self point:updatedPos];
            [self setContentOffset:scrollView point:oldPosition];
        }
    }
}

- (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary<NSKeyValueChangeKey,id> *)change context:(void *)context {
    if (!self.observing) {
        return;
    }
    if (![object isKindOfClass:[UIScrollView class]]) {
        return;
    }
    UIScrollView *scrollView = object;
    if (scrollView == self) {
        return;
    }
    NSValue *newValue = [change valueForKey:NSKeyValueChangeNewKey];
    NSValue *oldValue = [change valueForKey:NSKeyValueChangeOldKey];
    if ([newValue isKindOfClass:[NSValue class]] && [oldValue isKindOfClass:[NSValue class]]) {
        CGPoint new = [newValue CGPointValue];
        CGPoint old = [oldValue CGPointValue];
        CGFloat diff = old.y - new.y;
        if (diff > 0.0) {
            [self handleScrollUp:scrollView change:diff oldPosition:old];
        } else {
            [self handleScrollDown:scrollView change:diff oldPosition:old];
        }
    }
}

- (void)setContentOffset:(UIScrollView *)scrollView point:(CGPoint) point {
    self.observing = NO;
    scrollView.contentOffset = point;
    self.observing = YES;
}

@end

File: SJSegmentedScrollView/Classes/SJSegmentedScrollView.swift
//
//  SJSegmentedScrollView.swift
//  SJSegmentedScrollView
//
//  Created by Subins Jose on 10/06/16.
//  Copyright © 2016 Subins Jose. All rights reserved.
//
//  Permission is hereby granted, free of charge, to any person obtaining a copy of this software and
//    associated documentation files (the "Software"), to deal in the Software without restriction,
//    including without limitation the rights to use, copy, modify, merge, publish, distribute,
//    sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is
//    furnished to do so, subject to the following conditions:
//
//  The above copyright notice and this permission notice shall be included in all copies or
//    substantial portions of the Software.
//
//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT
//  LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
//  IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
//  WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
//  SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

import UIKit

class SJSegmentedScrollView: UIScrollView {
    
    var segmentView: SJSegmentView?
    
    var headerViewHeight: CGFloat! = 0 {
        didSet {
            headerHeightConstraint?.constant = headerViewHeight
        }
    }
    
    var segmentViewHeight: CGFloat! = 0 {
        didSet {
            segmentViewHeightConstraint?.constant = segmentViewHeight
        }
    }
    
    var headerViewOffsetHeight: CGFloat! = 0
    
    var selectedSegmentViewColor: UIColor! = UIColor.red {
        didSet {
            segmentView?.selectedSegmentViewColor = selectedSegmentViewColor
        }
    }
    
    var selectedSegmentViewHeight: CGFloat! = 0
    
    var segmentBounces = false
    
    var segmentTitleColor: UIColor! = UIColor.red {
        didSet {
            segmentView?.titleColor = segmentTitleColor
        }
    }
    
    var segmentSelectedTitleColor: UIColor? {
        didSet {
            segmentView?.selectedTitleColor = segmentSelectedTitleColor
        }
    }
    
    var segmentBackgroundColor: UIColor? {
        didSet {
            segmentView?.segmentBackgroundColor = segmentBackgroundColor
        }
    }
    
    var segmentShadow: SJShadow? {
        didSet {
            segmentView?.shadow = segmentShadow
        }
    }
    
    var segmentTitleFont: UIFont! = UIFont.systemFont(ofSize: 12) {
        didSet {
            segmentView?.font = segmentTitleFont
        }
    }
    
    var topSpacing: CGFloat?
    
    var bottomSpacing: CGFloat?
    
    var observing = true
    
    var headerView: UIView?
    
    var contentControllers: [UIViewController]?
    
    var contentViews = [UIView]()
    
    var contentView: SJContentView?
    
    var scrollContentView: UIView!
    
    var contentViewHeightConstraint: NSLayoutConstraint!
    
    var didSelectSegmentAtIndex: DidSelectSegmentAtIndex?
    
	var sjShowsVerticalScrollIndicator: Bool = false {
		didSet {
			showsVerticalScrollIndicator = sjShowsVerticalScrollIndicator
			contentView?.showsVerticalScrollIndicator = sjShowsVerticalScrollIndicator
		}
	}
    
	var sjShowsHorizontalScrollIndicator: Bool = false  {
		didSet {
			showsHorizontalScrollIndicator = sjShowsHorizontalScrollIndicator
			contentView?.showsHorizontalScrollIndicator = sjShowsHorizontalScrollIndicator
		}
	}
    
    private var viewObservers = [UIView]()
    var sjDisableScrollOnContentView: Bool = false {
        didSet {
            contentView?.isScrollEnabled = !sjDisableScrollOnContentView
        }
    }

    override init(frame: CGRect) {
        super.init(frame: frame)
        
        sizeToFit()
        translatesAutoresizingMaskIntoConstraints = false
        showsHorizontalScrollIndicator = sjShowsHorizontalScrollIndicator
        showsVerticalScrollIndicator = sjShowsVerticalScrollIndicator
        decelerationRate = .fast
        bounces = false
        
        addObserver(self, forKeyPath: "contentOffset",
                         options: [NSKeyValueObservingOptions.new, NSKeyValueObservingOptions.old],
                         context: nil)
        
        
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    deinit {
        removeObserver(self,
                            forKeyPath: "contentOffset",
                            context: nil)
        
        for view in viewObservers {
            view.removeObserver(self,
                                forKeyPath: "contentOffset",
                                context: nil)
        }
    }
    
    func setContentView() {
        
        if scrollContentView == nil {
            
            scrollContentView = UIView()
            scrollContentView.translatesAutoresizingMaskIntoConstraints = false
            addSubview(scrollContentView)
            
            let horizontalConstraints = NSLayoutConstraint.constraints(withVisualFormat: "H:|[contentView(==mainView)]|",
                                                                                       options: [],
                                                                                       metrics: nil,
                                                                                       views: ["contentView": scrollContentView, "mainView": self])
            addConstraints(horizontalConstraints)
            
            let contentHeight = getContentHeight()
            let verticalConstraints = NSLayoutConstraint.constraints(withVisualFormat: "V:|[contentView]|",
                                                                                     options: [],
                                                                                     metrics: nil,
                                                                                     views: ["contentView": scrollContentView])
            addConstraints(verticalConstraints)
            
            contentViewHeightConstraint = NSLayoutConstraint(item: scrollContentView,
                                                             attribute: .height,
                                                             relatedBy: .equal,
                                                             toItem: nil,
                                                             attribute: .notAnAttribute,
                                                             multiplier: 1.0,
                                                             constant: contentHeight)
            addConstraint(contentViewHeightConstraint)
        }
    }
    
    private var headerHeightConstraint: NSLayoutConstraint?
    
    func addHeaderView(_ headerView: UIView?) {
        
        if headerView != nil {
            
            self.headerView = headerView
            headerView?.translatesAutoresizingMaskIntoConstraints = false
            scrollContentView.addSubview(headerView!)
            
            let horizontalConstraints = NSLayoutConstraint.constraints(withVisualFormat: "H:|-0-[headerView]-0-|",
                                                                                       options: [],
                                                                                       metrics: nil,
                                                                                       views: ["headerView": headerView!])
            scrollContentView.addConstraints(horizontalConstraints)
            
            let verticalConstraints = NSLayoutConstraint.constraints(withVisualFormat: "V:|[headerView(\(headerViewHeight!))]",
                                                                                     options: [],
                                                                                     metrics: nil,
                                                                                     views: ["headerView": headerView!])
            scrollContentView.addConstraints(verticalConstraints)
            
            headerHeightConstraint = verticalConstraints[1]
        } else {
            
            headerViewHeight = headerViewOffsetHeight
        }
    }
    
    func addObserverFor(_ view: UIView) {
        viewObservers.append(view)
        view.addObserver(self, forKeyPath: "contentOffset",
                         options: [NSKeyValueObservingOptions.new, NSKeyValueObservingOptions.old],
                         context: nil)
    }
    
    func addContentView(_ contentView: UIView, frame: CGRect) {
        
        if self.contentView == nil {
            self.contentView = createContentView()
        }
        
        contentViews.append(contentView)
        self.contentView?.addContentView(contentView, frame: frame)
        self.contentView!.didSelectSegmentAtIndex = {[unowned self]
            (segment, index, animated) in
            self.didSelectSegmentAtIndex?(self.segmentView!.segments[index], index, animated)
        }
    }
    
    func updateSubviewsFrame(_ frame: CGRect) {
        
        contentViewHeightConstraint.constant = getContentHeight()
        contentView?.layoutIfNeeded()
        
        segmentView?.didChangeParentViewFrame(frame)
        contentView?.updateContentControllersFrame(frame)
    }
    
    //MARK: Private Functions
    func getContentHeight() -> CGFloat {
		
        var contentHeight = (superview?.bounds.height)! + headerViewHeight!
        contentHeight -= (topSpacing! + bottomSpacing! + headerViewOffsetHeight!)
        return contentHeight
    }
    
    private var segmentViewHeightConstraint: NSLayoutConstraint?
    
    func addSegmentView(_ controllers: [UIViewController], frame: CGRect) {
        
        if controllers.count > 1 {
            
            segmentView = SJSegmentView(frame: CGRect.zero)
			segmentView?.controllers					= controllers
            segmentView?.selectedSegmentViewColor		= selectedSegmentViewColor
            segmentView?.selectedSegmentViewHeight		= selectedSegmentViewHeight!
            segmentView?.titleColor						= segmentTitleColor
            segmentView?.selectedTitleColor             = segmentSelectedTitleColor
            segmentView?.segmentBackgroundColor			= segmentBackgroundColor
            segmentView?.font							= segmentTitleFont!
            segmentView?.shadow							= segmentShadow
            segmentView?.font							= segmentTitleFont!
            segmentView?.bounces						= false
            segmentView!.translatesAutoresizingMaskIntoConstraints = false
            segmentView!.didSelectSegmentAtIndex = {[unowned self]
                (segment, index, animated) in
                self.contentView?.movePageToIndex(index, animated: animated)
                self.didSelectSegmentAtIndex?(segment, index, animated)
            }
            
            segmentView?.setSegmentsView(frame)
            addSubview(segmentView!)
            
            let horizontalConstraints = NSLayoutConstraint.constraints(withVisualFormat: "H:|-0-[segmentView]-0-|",
                                                                                       options: [],
                                                                                       metrics: nil,
                                                                                       views: ["segmentView": segmentView!])
            addConstraints(horizontalConstraints)
            
            let view = headerView == nil ? self : headerView
            let verticalConstraints = NSLayoutConstraint.constraints(withVisualFormat: "V:[headerView]-0-[segmentView(\(segmentViewHeight!))]",
                                                                                     options: [],
                                                                                     metrics: nil,
                                                                                     views: ["headerView": view!,
                                                                                        "segmentView": segmentView!])
            addConstraints(verticalConstraints)
            
            segmentViewHeightConstraint = verticalConstraints[1]
        } else {
            
            segmentViewHeight = 0.0
        }
    }
    
    func addSegmentsForContentViews(_ titles: [String]) {
        
        let frame = CGRect(x: 0, y: headerViewHeight!,
                           width: bounds.size.width, height: segmentViewHeight!)
        segmentView = SJSegmentView(frame: frame)
        segmentView!.didSelectSegmentAtIndex = {
            [unowned self] (segment, index, animated) in
            self.contentView?.movePageToIndex(index, animated: animated)
        }
        addSubview(segmentView!)
    }
    
    func createContentView() -> SJContentView {
        
        let contentView = SJContentView(frame: CGRect.zero)
		contentView.showsVerticalScrollIndicator = sjShowsVerticalScrollIndicator
		contentView.showsHorizontalScrollIndicator = sjShowsHorizontalScrollIndicator
        contentView.isScrollEnabled = !sjDisableScrollOnContentView
        contentView.translatesAutoresizingMaskIntoConstraints = false
		contentView.bounces = segmentBounces
        scrollContentView.addSubview(contentView)
        
        let horizontalConstraints = NSLayoutConstraint.constraints(withVisualFormat: "H:|-0-[contentView]-0-|",
                                                                                   options: [],
                                                                                   metrics: nil,
                                                                                   views: ["contentView": contentView])
        scrollContentView.addConstraints(horizontalConstraints)
        
        let verticalConstraints = NSLayoutConstraint.constraints(withVisualFormat: "V:[headerView]-\(segmentViewHeight!)-[contentView]-0-|",
                                                                                 options: [],
                                                                                 metrics: nil,
                                                                                 views: ["headerView": headerView!,
                                                                                    "contentView": contentView])

        
        scrollContentView.addConstraints(verticalConstraints)
        return contentView
    }
    
    func handleScrollUp(_ scrollView: UIScrollView,
                        change: CGFloat,
                        oldPosition: CGPoint) {

		if headerViewHeight != 0.0 && contentOffset.y != 0.0 {
			if scrollView.contentOffset.y < 0.0 {
				if contentOffset.y >= 0.0 {

					var yPos = contentOffset.y - change
					yPos = yPos < 0 ? 0 : yPos
					let updatedPos = CGPoint(x: contentOffset.x, y: yPos)
					setContentOffset(self, point: updatedPos)
					setContentOffset(scrollView, point: oldPosition)
				}
			}
		}
    }

    func handleScrollDown(_ scrollView: UIScrollView,
                          change: CGFloat,
                          oldPosition: CGPoint) {
        
        let offset = (headerViewHeight! - headerViewOffsetHeight!)
        
        if contentOffset.y < offset {
            
            if scrollView.contentOffset.y >= 0.0 {
                
                var yPos = contentOffset.y - change
                yPos = yPos > offset ? offset : yPos
                let updatedPos = CGPoint(x: contentOffset.x, y: yPos)
                setContentOffset(self, point: updatedPos)
                setContentOffset(scrollView, point: oldPosition)
            }
        }
    }

	override func observeValue(forKeyPath keyPath: String?,
	                           of object: Any?,
	                           change: [NSKeyValueChangeKey : Any]?,
	                           context: UnsafeMutableRawPointer?) {
		if !observing { return }

		let scrollView = object as? UIScrollView
		if scrollView == nil { return }
		if scrollView == self { return }

		let changeValues = change! as [NSKeyValueChangeKey: AnyObject]

		if let new = changeValues[NSKeyValueChangeKey.newKey]?.cgPointValue,
			let old = changeValues[NSKeyValueChangeKey.oldKey]?.cgPointValue {

			let diff = old.y - new.y

			if diff > 0.0 {

				handleScrollUp(scrollView!,
				                    change: diff,
				                    oldPosition: old)
			} else {

				handleScrollDown(scrollView!,
				                      change: diff,
				                      oldPosition: old)
			}
		}
	}

    func setContentOffset(_ scrollView: UIScrollView, point: CGPoint) {
        observing = false
        scrollView.contentOffset = point
        observing = true
    }
}
